### DISCUSSION TITLE
berry rule with wildcard

### DISCUSSION THREAD

**Discussion by DetlefR** [arendst/Tasmota#18766] [Q&A] [answered]
Hello,

I'm testing berry scripting with DS18B20 sensors. I use rules to process the sensor values.
Like 
tasmota.add_rule('DS18B20',s_ds18b20)
It is working well with one sensor. With two, I must declare two rules like
tasmota.add_rule('DS18B20-1',s_ds18b20)
tasmota.add_rule('DS18B20-2',s_ds18b20)

Is there a way to use wildcards to use one rule for all (unknown how much) sensor(s).

Best regards
Detlef

**Comment by sfromis** [comment, marked as answer]
You need only one rule, for the first sensor, even with multiple connected. Minimal example rule to show what I mean:
`tasmota.add_rule("DS18B20-1",/value,topic,data -> print(data.find('DS18B20-2',{}).find('Temperature',nil)))`
The third argument `data` is a map with all the sensor values, where you can retrieve individual values.

Obviously, instead of just the print closure, you can have a function with code to do the sensor handling.

The 2nd argument to the `find` function calls is a default value, to prevent the code from crashing in case of values not being present for whatever (hopefully unlikely) reason.

The only wildcard available would be `tasmota.add_rule('?', .....` firing on all possible events, which would really be a bit too much.

**Comment by DetlefR** [comment]
Hello,

thank you for the idea. 
I have solved it with an iteration over all possible thermometer.
```
def s_ds18b20(wert,trigger,msg)
 for t: msg.keys()
  print('DS18B20',t,msg[t]['Id'],msg[t]['Temperature'])
  .
  .
  . 
 end 
end

tasmota.add_rule('DS18B20',s_ds18b20) 
tasmota.add_rule('DS18B20-1',s_ds18b20) 
```

**Reply by sfromis** [reply]
Beware that your function will fail if having other sensors than just DS18B20 present. Referencing a map via `[]` with a key not found will throw an exception. One way of avoiding that is using `find` as in my example. Using `if t.contains('Id')` would also be a strategy. Or checking if the keys you get starts with DS18B20.

**Comment by DetlefR** [comment]
You are right.
I have not considered another sensors yet.
```
def s_ds18b20(wert,trigger,msg)
 for t: msg.keys()
  if string.find(t,'DS18B20') >= 0
   print('DS18B20',t,msg[t]['Id'],msg[t]['Temperature'])
   .
   .
  . 
  end 
 end 
end
```
